packages/frontend/workspace/src/manager.ts

```diff
@@ -0,0 +1,189 @@
+import { DebugLogger } from '@affine/debug';
+import { WorkspaceFlavour } from '@affine/env/workspace';
+import { assertEquals } from '@blocksuite/global/utils';
+import type { Workspace as BlockSuiteWorkspace } from '@blocksuite/store';
+import {
+  fixWorkspaceVersion,
+  guidCompatibilityFix,
+} from '@toeverything/infra/blocksuite';
+import { applyUpdate, encodeStateAsUpdate } from 'yjs';
+
+import type { BlobStorage } from '.';
+import type { WorkspaceFactory } from './factory';
+import { LOCAL_WORKSPACE_LOCAL_STORAGE_KEY } from './impl/local/consts';
+import type { WorkspaceList } from './list';
+import type { WorkspaceMetadata } from './metadata';
+import { WorkspacePool } from './pool';
+import type { Workspace } from './workspace';
+
+const logger = new DebugLogger('affine:workspace-manager');
+
+/**
+ * # `WorkspaceManager`
+ *
+ * This class acts as the central hub for managing various aspects of workspaces.
+ * It is structured as follows:
+ *
+ * ```
+ *                ┌───────────┐
+ *                │ Workspace │
+ *                │  Manager  │
+ *                └─────┬─────┘
+ *        ┌─────────────┼─────────────┐
+ *    ┌───┴───┐     ┌───┴───┐   ┌─────┴─────┐
+ *    │ List  │     │ Pool  │   │ Factories │
+ *    └───────┘     └───────┘   └───────────┘
+ * ```
+ *
+ * Manage every about workspace
+ *
+ * # List
+ *
+ * The `WorkspaceList` component stores metadata for all workspaces, also include workspace avatar and custom name.
+ *
+ * # Factories
+ *
+ * This class contains a collection of `WorkspaceFactory`,
+ * We utilize `metadata.flavour` to identify the appropriate factory for opening a workspace.
+ * Once opened, workspaces are stored in the `WorkspacePool`.
+ *
+ * # Pool
+ *
+ * The `WorkspacePool` use reference counting to manage active workspaces.
+ * Calling `use()` to create a reference to the workspace. Calling `release()` to release the reference.
+ * When the reference count is 0, it will close the workspace.
+ *
+ */
+export class WorkspaceManager {
```

Comment:
Good question, Right now, we update React using some onChanged events.

Jotai isn't really a good fit in this case. To handle multiple instances, the atom needs to be part of the class.

Also, the atom doesn't hold data by itself.

For it to work, the class needs a store. But it's a bit odd for a class to have a Jotai store or get a Jotai store from global.

I think we need an atom that comes with its own store.

Reply:
One issue is how to make a top-down data flow pattern to let the biz modules read/write the properties, e.g., a page's meta is derived all the way from the root manager.

Regarding the implementation, I am fine to have Jotai but it seems the class itself will have little control over the data held by Jotai; in our case a simple `useSyncExternalStore` could bridge global state stored in manager with hooks that can be consumed by the react side.
Another solution may be that reversing the relation between Jotai and the manager. Instead of using manager to control atoms, but define atoms outside of the manager that consumes the manager's properties and listen to change events.

What do you think?